File: shared/src/commonMain/kotlin/org/nihongo/mochi/ui/grammar/GrammarQuizScreen.kt
package org.nihongo.mochi.ui.grammar

import androidx.activity.compose.BackHandler
import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import org.jetbrains.compose.resources.stringResource
import org.nihongo.mochi.domain.grammar.ExercisePayload
import org.nihongo.mochi.domain.models.AnswerButtonState
import org.nihongo.mochi.domain.models.GameState
import org.nihongo.mochi.presentation.MochiBackground
import org.nihongo.mochi.shared.generated.resources.Res
import org.nihongo.mochi.ui.components.GameAnswerButton
import org.nihongo.mochi.ui.components.GameProgressBar
import org.nihongo.mochi.ui.components.GameQuestionCard
import org.nihongo.mochi.ui.components.ExitConfirmationDialog
import org.nihongo.mochi.ui.components.GameResultOverlay
import org.nihongo.mochi.ui.theme.AppTheme
import org.nihongo.mochi.shared.generated.resources.*

@Composable
fun GrammarQuizScreen(
    viewModel: GrammarQuizViewModel,
    onBackClick: () -> Unit
) {
    val state by viewModel.state.collectAsState()
    var showExitDialog by remember { mutableStateOf(false) }
    var showResultOverlay by remember { mutableStateOf(false) }

    BackHandler(enabled = state.gameState != GameState.Finished && !showResultOverlay) {
        showExitDialog = true
    }

    LaunchedEffect(state.gameState) {
        if (state.gameState == GameState.Finished) {
            showResultOverlay = true
        }
    }

    AppTheme {
        MochiBackground {
            Scaffold(
                containerColor = Color.Transparent,
                topBar = {
                    Row(
                        modifier = Modifier.fillMaxWidth().padding(8.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        GameProgressBar(
                            statuses = state.progressHistory,
                            maxItems = 10
                        )
                    }
                }
            ) { paddingValues ->
                Box(modifier = Modifier.fillMaxSize().padding(paddingValues)) {
                    if (state.gameState == GameState.Loading) {
                        CircularProgressIndicator(modifier = Modifier.align(Alignment.Center))
                    } else {
                        QuizContent(state, viewModel)
                    }
                }

                if (showExitDialog) {
                    ExitConfirmationDialog(
                        onConfirm = { 
                            showExitDialog = false
                            showResultOverlay = true 
                        },
                        onDismiss = { showExitDialog = false },
                        onPause = { },
                        onResume = { }
                    )
                }

                if (showResultOverlay) {
                    GameResultOverlay(
                        isVictory = state.gameState == GameState.Finished,
                        score = "${(state.globalMasteryPercent * 100).toInt()}%",
                        stats = listOf(
                            stringResource(Res.string.game_result_title_session)  to "${(state.sessionMasteryPercent * 100).toInt()}%",
                            stringResource(Res.string.game_result_title_global) to "${(state.globalMasteryPercent * 100).toInt()}%"
                        ),
                        title = stringResource(Res.string.game_result_lot_mastery_grammar),
                        onReplayClick = {
                            showResultOverlay = false
                            viewModel.replay()
                        },
                        onMenuClick = {
                            showResultOverlay = false
                            onBackClick()
                        }
                    )
                }
            }
        }
    }
}

@Composable
fun QuizContent(state: GrammarQuizState, viewModel: GrammarQuizViewModel) {
    val payload = state.currentExercisePayload ?: return

    Column(
        modifier = Modifier.fillMaxSize()
    ) {
        // 1/ Question area centered vertically with fixed size card
        Box(
            modifier = Modifier
                .weight(1f)
                .fillMaxWidth(),
            contentAlignment = Alignment.Center
        ) {
            val questionText = when (payload) {
                is ExercisePayload.FillBlank -> payload.sentence.replace("__", " ___★___ ")
                is ExercisePayload.Underline -> payload.sentence.replace("[", "【").replace("]", "】")
                is ExercisePayload.Paraphrase -> payload.baseSentence
                is ExercisePayload.WordUsage -> "「${payload.word}」の使い方が正しいものを選んでください。"
                is ExercisePayload.SentenceOrder -> {
                    val holes = List(payload.blocks.size) { i -> if (i == state.currentStarIndex) " ★ " else " ___ " }
                    "${payload.prefix} ${holes.joinToString("")} ${payload.suffix}"
                }
            }

            GameQuestionCard(
                text = questionText,
                fontSize = if (payload is ExercisePayload.WordUsage || payload is ExercisePayload.Paraphrase) 18.sp else 22.sp,
                modifier = Modifier.size(width = 340.dp, height = 240.dp)
            )
        }

        // 2/ Answer buttons area at the bottom
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(12.dp)
        ) {
            state.currentOptions.chunked(2).forEach { row ->
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    row.forEach { option ->
                        val isSelected = state.selectedOption == option
                        val isActuallyCorrect = checkIsCorrect(option, payload, state.currentStarIndex)
                        
                        val buttonState = when {
                            isSelected && state.isAnswerCorrect == true -> AnswerButtonState.CORRECT
                            isSelected && state.isAnswerCorrect == false -> AnswerButtonState.INCORRECT
                            state.selectedOption != null && isActuallyCorrect -> AnswerButtonState.CORRECT
                            else -> AnswerButtonState.DEFAULT
                        }
                        
                        GameAnswerButton(
                            text = option,
                            state = buttonState,
                            enabled = state.selectedOption == null,
                            modifier = Modifier.weight(1f).height(100.dp),
                            fontSizeSp = if (payload is ExercisePayload.WordUsage || payload is ExercisePayload.Paraphrase) 14 else 18,
                            onClick = { viewModel.onOptionSelected(option) }
                        )
                    }
                    if (row.size == 1) {
                        Spacer(modifier = Modifier.weight(1f))
                    }
                }
                Spacer(modifier = Modifier.height(8.dp))
            }
        }
    }
}

private fun checkIsCorrect(option: String, payload: ExercisePayload, starIndex: Int): Boolean {
    return when (payload) {
        is ExercisePayload.FillBlank -> option == payload.correct
        is ExercisePayload.Underline -> option == payload.correct
        is ExercisePayload.Paraphrase -> option == payload.correct
        is ExercisePayload.WordUsage -> payload.options.find { it.text == option }?.is_correct ?: false
        is ExercisePayload.SentenceOrder -> {
            if (starIndex < payload.blocks.size) {
                option == payload.blocks[starIndex]
            } else false
        }
    }
}

File: flutter_app/lib/grammar_quiz_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'models/grammar_quiz.dart';
import 'models/quiz_models.dart';
import 'providers/grammar_quiz_provider.dart';
import 'providers/settings_provider.dart';
import 'repositories/exercise_repository.dart';
import 'repositories/score_repository.dart';
import 'services/audio_service.dart';
import 'widgets/mochi_background.dart';
import 'widgets/game_components.dart';

class GrammarQuizScreen extends StatelessWidget {
  final List<String> grammarTags;

  const GrammarQuizScreen({super.key, required this.grammarTags});

  @override
  Widget build(BuildContext context) {
    return ChangeNotifierProvider(
      create: (context) {
        final provider = GrammarQuizProvider(
          exerciseRepo: context.read<ExerciseRepository>(),
          scoreRepo: context.read<ScoreRepository>(),
          audioService: context.read<AudioService>(),
          grammarTags: grammarTags,
        );
        provider.startQuiz();
        return provider;
      },
      child: const GrammarQuizView(),
    );
  }
}

class GrammarQuizView extends StatelessWidget {
  const GrammarQuizView({super.key});

  @override
  Widget build(BuildContext context) {
    final provider = context.watch<GrammarQuizProvider>();

    if (provider.gameState == GameState.loading) {
      return const Scaffold(body: Center(child: CircularProgressIndicator()));
    }

    if (provider.gameState == GameState.finished) {
      return _buildFinishedScreen(context, provider);
    }

    final payload = provider.currentPayload;

    return PopScope(
      canPop: false,
      onPopInvokedWithResult: (didPop, result) {
        if (didPop) return;
        _showExitDialog(context, provider);
      },
      child: Scaffold(
        appBar: AppBar(
          title: GameProgressBar(
            statuses: provider.progressHistory,
            maxItems: 10,
          ),
          backgroundColor: Colors.transparent,
          elevation: 0,
          automaticallyImplyLeading: false,
          leading: IconButton(
            icon: const Icon(Icons.arrow_back),
            onPressed: () => _showExitDialog(context, provider),
          ),
        ),
        backgroundColor: Colors.transparent,
        extendBodyBehindAppBar: true,
        body: MochiBackground(
          child: SafeArea(
            child: Column(
              children: [
                Expanded(
                  child: Center(
                    child: SingleChildScrollView(
                      padding: const EdgeInsets.symmetric(horizontal: 24, vertical: 8),
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          if (payload != null)
                            GameQuestionCard(
                              text: _getQuestionText(payload, provider.currentStarIndex),
                              fontSize: (payload is WordUsagePayload) ? 18 : 22,
                            ),
                          const SizedBox(height: 32),
                          if (payload != null)
                            _buildOptionsArea(context, provider, payload),
                        ],
                      ),
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  void _showExitDialog(BuildContext context, GrammarQuizProvider provider) {
    showDialog(
      context: context,
      barrierDismissible: false,
      builder: (dialogContext) => ExitConfirmationDialog(
        onConfirm: () {
          Navigator.pop(dialogContext); // close dialog
          Navigator.pop(context); // exit quiz
        },
        onDismiss: () => Navigator.pop(dialogContext), // resume
      ),
    );
  }

  Widget _buildFinishedScreen(BuildContext context, GrammarQuizProvider provider) {
    final settings = context.read<SettingsProvider>();
    return GameResultOverlay(
      isVictory: provider.quizScore >= 8,
      title: settings.getString("game_result_lot_mastery_grammar"),
      score: "${(provider.globalMasteryPercent * 100).toInt()}%",
      stats: [
        MapEntry(settings.getString("game_result_title_session"), "${(provider.sessionMasteryPercent * 100).toInt()}%"),
        MapEntry(settings.getString("game_result_title_global"), "${(provider.globalMasteryPercent * 100).toInt()}%"),
      ],
      onReplayClick: () => provider.replay(),
      onMenuClick: () => Navigator.pop(context),
    );
  }

  String _getQuestionText(ExercisePayload payload, int starIndex) {
    if (payload is FillBlankPayload) {
      return payload.sentence.replaceAll("__", " ___★___ ");
    } else if (payload is UnderlinePayload) {
      return payload.sentence.replaceAll("[", "【").replaceAll("]", "】");
    } else if (payload is WordUsagePayload) {
      return "「${payload.word}」の使い方が正しいものを選んでください。";
    } else if (payload is SentenceOrderPayload) {
      final holes = List.generate(payload.blocks.length, (i) => i == starIndex ? " ★ " : " ___ ").join("");
      return "${payload.prefix} $holes ${payload.suffix}";
    }
    return "";
  }

  Widget _buildOptionsArea(BuildContext context, GrammarQuizProvider provider, ExercisePayload payload) {
    final options = provider.currentOptions;
    final List<List<String>> chunkedOptions = [];
    for (int i = 0; i < options.length; i += 2) {
      chunkedOptions.add(options.sublist(i, i + 2 > options.length ? options.length : i + 2));
    }

    return Column(
      children: chunkedOptions.map((row) {
        return Padding(
          padding: const EdgeInsets.only(bottom: 8.0),
          child: Row(
            children: row.map((option) {
              final isSelected = provider.selectedOption == option;
              final isActuallyCorrect = _isOptionCorrect(option, payload);
              
              AnswerButtonState buttonState = AnswerButtonState.defaultState;
              if (isSelected && (provider.isAnswerCorrect ?? false)) {
                buttonState = AnswerButtonState.correct;
              } else if (isSelected && !(provider.isAnswerCorrect ?? false)) {
                buttonState = AnswerButtonState.incorrect;
              } else if (provider.selectedOption != null && isActuallyCorrect) {
                buttonState = AnswerButtonState.correct;
              }

              return Expanded(
                child: Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 4.0),
                  child: GameAnswerButton(
                    text: option,
                    state: buttonState,
                    enabled: provider.selectedOption == null,
                    fontSize: (payload is WordUsagePayload) ? 14 : 18,
                    onClick: () => provider.submitAnswer(option),
                  ),
                ),
              );
            }).toList(),
          ),
        );
      }).toList(),
    );
  }

  bool _isOptionCorrect(String option, ExercisePayload p) {
    if (p is FillBlankPayload) return option == p.correct;
    if (p is UnderlinePayload) return option == p.correct;
    if (p is WordUsagePayload) {
      final opt = p.options.firstWhere(
        (o) => o.text == option,
        orElse: () => UsageOption(text: "", isCorrect: false),
      );
      return opt.isCorrect;
    }
    if (p is SentenceOrderPayload) return option == p.blocks.first;
    return false;
  }
}

class GameQuestionCard extends StatelessWidget {
  final String text;
  final double fontSize;

  const GameQuestionCard({
    super.key,
    required this.text,
    required this.fontSize,
  });

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    return Card(
      elevation: 12,
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(24)),
      child: Container(
        width: 340,
        height: 240,
        padding: const EdgeInsets.all(24),
        alignment: Alignment.center,
        decoration: BoxDecoration(
          color: theme.colorScheme.surface,
          borderRadius: BorderRadius.circular(24),
        ),
        child: SingleChildScrollView(
          child: Text(
            text,
            textAlign: TextAlign.center,
            style: TextStyle(
              fontSize: fontSize,
              fontWeight: FontWeight.bold,
              color: theme.colorScheme.onSurface,
            ),
          ),
        ),
      ),
    );
  }
}
